codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/requestcompression/RequestCompression.java

```diff
@@ -0,0 +1,160 @@
+                        return;
+                    }
+                    SymbolProvider symbolProvider = GoCodegenPlugin.createSymbolProvider(model, settings);
+                    String funcName = getAddRequestCompressionMiddlewareFuncName(
+                            symbolProvider.toSymbol(operation).getName()
+                    );
+                    runtimeClientPlugins.add(RuntimeClientPlugin.builder().operationPredicate((m, s, o) -> {
+                        if (!o.hasTrait(RequestCompressionTrait.class)) {
+                            return false;
+                        }
+                        return o.equals(operation);
+                    }).registerMiddleware(MiddlewareRegistrar.builder()
+                        .resolvedFunction(SymbolUtils.buildPackageSymbol(funcName))
+                        .useClientOptions().build())
+                        .build());
+                });
+    }
+
+    @Override
+    public void writeAdditionalFiles(
+            GoSettings settings,
+            Model model,
+            SymbolProvider symbolProvider,
+            GoDelegator goDelegator
+    ) {
+        ServiceShape service = settings.getService(model);
+        for (ShapeId operationID : service.getAllOperations()) {
+            OperationShape operation = model.expectShape(operationID, OperationShape.class);
+            if (!operation.hasTrait(RequestCompressionTrait.class)) {
+                continue;
+            }
+            goDelegator.useShapeWriter(operation, writeMiddlewareHelper(symbolProvider, operation));
+        }
+    }
+
+
+    public static boolean isRequestCompressionService(Model model, ServiceShape service) {
+        return TopDownIndex.of(model)
+                .getContainedOperations(service).stream()
+                .anyMatch(it -> it.hasTrait(RequestCompressionTrait.class));
+    }
+
+    private GoWriter.Writable writeMiddlewareHelper(SymbolProvider symbolProvider, OperationShape operation) {
+        String operationName = symbolProvider.toSymbol(operation).getName();
+        RequestCompressionTrait trait = operation.expectTrait(RequestCompressionTrait.class);
+
+        return goTemplate("""
+                func $add:L(stack $stack:P, options Options) error {
+                    return $addInternal:T(stack, options.DisableRequestCompression, options.RequestMinCompressSizeBytes,
+                    $algorithms:W)
+                }
+                """,
+                MapUtils.of(
+                "add", getAddRequestCompressionMiddlewareFuncName(operationName),
+                "stack", SmithyGoTypes.Middleware.Stack,
+                "addInternal", SmithyGoTypes.Private.RequestCompression.AddRequestCompression,
+                "algorithms", generateAlgorithmList(trait.getEncodings())
+                ));
+    }
+
+    @Override
+    public List<RuntimeClientPlugin> getClientPlugins() {
+       runtimeClientPlugins.add(
```

Comment:
Use `List.of()` instead, and register all client plugins here immutably instead of using state.

Reply:
`getClientPlugins` should probably accept the `Model` and service shape ID much like these other functions do long-term.